core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncPathCache.java

```diff
@@ -11,201 +11,348 @@
+    } else if (lastInvalidationTime == 0 && intervalMs < 0) {
+      // if no invalidation, and a negative interval then do not sync
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    } else if (lastInvalidationTime >= lastValidationTime) {
+      // if a recent invalidation then we always sync
+      result = SyncCheck.shouldSyncWithTime(lastValidationTime);
+    } else if (intervalMs < 0) {
+      // do not sync with a negative interval
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    } else if ((currentTime - lastValidationTime) >= intervalMs) {
+      // syncing is needed based on an interval
+      result = SyncCheck.shouldSyncWithTime(lastValidationTime);
+    } else {
+      // syncing is not needed based on an interval
+      result = SyncCheck.shouldNotSyncWithTime(lastValidationTime);
+    }
+    LOG.debug("Result of should sync path {}: {}, invalidation time: {}, validation time: {},"
+            + " clock time: {}",
+        result, path, lastInvalidationTime, lastValidationTime, currentTime);
+    return result;
   }
 
   /**
-   * Determines if the sync should be performed.
-   *
-   * @param syncTime the {@link SyncTime} to examine
-   * @param intervalMs the sync interval, in ms
-   * @param checkRecursive checks the recursive sync time if true, checks the standard sync time
-   *                       otherwise
-   * @return true if the sync should be performed
+   * Notify that a path has been invalidated.
+   * @param path the path
    */
-  private boolean shouldSyncInternal(@Nullable SyncTime syncTime, long intervalMs,
-      boolean checkRecursive) {
-    if (syncTime == null) {
-      return true;
-    }
-    long lastSyncMs = syncTime.getLastSyncMs();
-    if (checkRecursive) {
-      lastSyncMs = syncTime.getLastRecursiveSyncMs();
+  @VisibleForTesting
+  public void notifyInvalidation(AlluxioURI path) throws InvalidPathException {
+    String currPath = path.getPath();
+    long time = mClock.millis();
+    notifyInvalidationInternal(currPath, time);
+  }
+
+  private void notifyInvalidationInternal(String currPath, long time)
+      throws InvalidPathException {
+    LOG.debug("Set sync invalidation for path {} at time {}", currPath, time);
+    int parentLevel = 0;
+
+    if (currPath.equals(AlluxioURI.SEPARATOR)) {
+      try (LockResource ignored = new LockResource(mRootLock)) {
+        mRoot.setInvalidationTime(time);
+      }
+    } else {
+      mItems.asMap().compute(currPath, (key, state) -> {
+        if (state == null) {
+          state = new SyncState(false);
+        }
+        // set the invalidation time for the path
+        state.setInvalidationTime(time);
+        return state;
+      });
     }
-    if (lastSyncMs == SyncTime.UNSYNCED) {
-      // was not synced ever, so should sync
-      return true;
+    while (!currPath.equals(AlluxioURI.SEPARATOR)) {
+      currPath = PathUtils.getParent(currPath);
```

Comment:
If you get a chance, in your stress test please keep an eye on the temp string usage, I'm a little concerned you will observe a non-trivial cost here. For such temp strings generated from split and substring, they typically get recycled in the Young gen in your JVM and you will observe much more Young GCs. If you observe you process, you may observe sth like below:
```
$ jstat -gcutil <alluxio master pid> 1000
Column S0, S1, E: You may see the numbers here going up very rapidly, because the eden and young gen space are devoured by a huge amount of temp strings. 
Column YGC, YGCT: You will then observe many Young GCs and the time here going up much faster than a cluster without these code change.
Column O, FGC, FGCT: Almost none of those temp strings are promoted to the old region so the O column consumption should not go up fast. There will also be very few Full GCs, not much difference from a cluster without these code change.
```
The young GCs are more often stop-the-world which introduce short pauses to your JVM. Typically those pauses are all right, despite that they hurt your JVM's total throughput. Modern JVMs are quite efficient in allocating and recycling from the young gen, so i'm not as concerned as allocating twice as much long-term objects. But still, the throughput drop might be non-trivial given your volume.

Reply:
I agree that the usage of paths and the class AlluxioURI in Alluxio needs to change. I think we can leave that to a future PR where a new design for handling paths is decided one.